### PR TITLE
Refactor moe codebase

### PR DESCRIPTION
**Pull request by kyuyeunk**
# Description

Cleanup fused moe file. No change in functionality.

- Add type annotations
- Add better function documentation
- Remove `transpose_rhs` since it's always `True`
- Remove unnecessary sharding constraints which were already enforced via `shard_map` 
- Replace `jax.experimental.shard_map.shard_map` into `jax.shard_map`
- Remove `fused_moe_func_padded` since padding is now handled by jax runner
- Remove unnecessary arguments that can be inferred from tensor or mesh shapes

# Tests

https://buildkite.com/tpu-commons/tpu-inference-ci/builds/6118

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/vllm/fused_moe.py [hunk 1]
```diff
@@ -23,7 +23,10 @@ def activation_fn(activation: str, x1, x2):
                 f"FusedMoE does not support {activation} activation")
 
 
-def _swigluoai(x1, x2, alpha=1.702, limit=7.0):
+def _swigluoai(x1: jax.Array,
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/fused_moe.py:25]: could you paste the test commands for the 3 files in this PR?

Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:25]: Sure! Here they are

```
pytest -v -s tests/layers/vllm/test_unquantized.py
pytest -v -s tests/layers/vllm/test_mxfp4.py
```

Also, as I have linked in the PR description, I have verified that CI all passes: https://buildkite.com/tpu-commons/tpu-inference-ci/builds/6118

File: tpu_inference/layers/vllm/fused_moe.py [hunk 2]
```diff
@@ -103,96 +105,93 @@ def tensor_sharded_gmm_merged_column_parallel(
...
+            group_offset=jnp.array(0),
+        )
 
     gmm_result = shard_map(
         _gmm,
         mesh=mesh,
         in_specs=(P("data", None), P(None, "model", None), P("data")),
         out_specs=(P("data", "model")),
-        check_rep=False,
+        check_vma=False,
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/fused_moe.py:129]: why set it to False? Can it be true?

File: tpu_inference/layers/vllm/fused_moe.py [hunk 3]
```diff
@@ -103,96 +105,93 @@ def tensor_sharded_gmm_merged_column_parallel(
...
             in_specs=(P("data", "model"), P(None, "model"), P("data")),
             out_specs=(P("data", "model")),
         )(gmm_result, rhs_bias, group_sizes)
+    gmm_result = gmm_result.astype(lhs.dtype)
 
-    n_shards = mesh.shape["model"]
-    output_sizes = [intermediate_size, intermediate_size]
-
+    tp_size = mesh.shape["model"]
+    output_sizes = [gmm_result.shape[-1] // 2] * 2
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/fused_moe.py]: i'd prefer the previous version which is much clearer `[intermediate_size, intermediate_size]`

File: tpu_inference/layers/vllm/fused_moe.py [hunk 4]
```diff
@@ -201,42 +200,33 @@ def _add_bias(gmm_result_local, rhs_bias_local, group_sizes_global):
...
 
-    return gmm_result
+    return gmm_result.astype(lhs.dtype)
 
 
 def expert_sharded_gmm(
     lhs: jax.Array,
     rhs: jax.Array,
     group_sizes: jax.Array,
-    transpose_rhs: bool,
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/fused_moe.py:211]: thanks for simplifying it!

File: tpu_inference/layers/vllm/fused_moe.py [hunk 5]
```diff
@@ -263,30 +253,20 @@ def _gmm(lhs, rhs, group_sizes, group_offset):
...
     # For i-th shard, it is responsible groups (AKA experts) from
     # i*num_experts_per_shard to (i+1)*num_experts_per_shard We sum them up to
     # get total rows in that shard, and that is the size for shard to send to
     # its peers. This is also the number of non-zero rows from the gmm results.
     # In the working example, send_sizes would be [3, 2, 5, 4]
-    send_sizes = jnp.array([
-        group_sizes[i * num_experts_per_shard:(i + 1) *
-                    num_experts_per_shard].sum() for i in range(ep_size)
-    ])
+    send_sizes = group_sizes.reshape(-1, num_experts_per_shard).sum()
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/fused_moe.py]: what's the shape of group_sizes before this line?

File: tpu_inference/layers/vllm/fused_moe.py (hunk 2 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:129]: so for context, when using `shard_map`, it is really easy to mess up sharding.

For example, if we have mesh `{"data": 4, "model" 1}` and `shard_map`'s `out_specs=P(None, "model")`. Correct sharding means the data is replicated along `"data"` dimension. but with `shard_map`, programmer might make a mistake where data is different along `"data"` dimension. When using `check_vma=True`, compiler runs a check during compile time and ensure this doesn't happen.

But when using kernel, this check does not work because compiler does not have an inside view of the kernel. So we disable it.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 5 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py]: it's `[num_total_tokens]` where `num_total_tokens=num_tokens_per_shard * num_experts_per_shard`. So this reshape transforms it into `[num_tokens_per_shard, num_experts_per_shard]`.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 3 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py]: done. as long as we dont have to add unnecessary argument and infer from existing input / mesh shapes, it's all fine by me.

File: tpu_inference/layers/vllm/fused_moe.py [hunk 6]
```diff
@@ -410,51 +409,41 @@ def _process_tokens_locally(hidden_states_local, topk_indices_local):
         mesh=mesh,
         in_specs=(P("data", None), P("data", None)),
         out_specs=(P("data", None), P("data"), P("data")),
-        check_rep=False,
     )(hidden_states, topk_indices)
+
     if use_ep:
         x = expert_sharded_gmm(
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/fused_moe.py:419]: Somehow I couldn't find a test that triggers this case (use_ep==True, line415). I tried "pytest tests/layers/vllm/test_unquantized.py -k test_fused_moe" but it didn't. Do you know which test exercise this case?

Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:419]: Here it is: https://github.com/vllm-project/tpu-inference/blob/77cb15682a4955f5ee5a90213b40e8885f930dc5/tests/layers/vllm/test_unquantized.py#L406

File: tpu_inference/layers/vllm/quantization/unquantized.py [hunk 7]
```diff
@@ -214,76 +216,56 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
 
-            # Reshape and transpose w13_weight to (num_experts, 2, hidden_size, intermediate_size)
             w13_reshaped = w13_weight.reshape(num_experts, 2,
                                               intermediate_size, hidden_size)
-            w13_weight_transposed = jnp.transpose(w13_reshaped, (0, 1, 3, 2))
 
-            # Transpose w2_weight to (num_experts, intermediate_size, hidden_size)
-            w2_weight_transposed = jnp.transpose(w2_weight, (0, 2, 1))
+            # Transpose non-constracting dim to right most dim
+            w13_weight_transposed = jnp.swapaxes(w13_reshaped, 2, 3)
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/quantization/unquantized.py:231]: previously it's transpose. Why is it changed to swapaxes? They should to be different

File: tpu_inference/layers/vllm/quantization/unquantized.py [hunk 8]
```diff
@@ -214,76 +216,56 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
-                w2_bias = jax.device_put(
-                    w2_bias,
-                    Format(Layout((0, 1)),
-                           NamedSharding(self.mesh, P("model", None))))
-
+                    w13_bias, Format(Layout((0, 1, 2)), ep_sharding))
+                w2_bias = jax.device_put(w2_bias,
+                                         Format(Layout((0, 1)), ep_sharding))
         else:
-            # Original logic for non-kernel path
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/quantization/unquantized.py:260]: could you leave the comment?

File: tpu_inference/layers/vllm/quantization/unquantized.py [hunk 9]
```diff
@@ -168,7 +168,7 @@ def __init__(self,
                  ep_axis_name: str = 'model'):
         super().__init__(moe)
         self.mesh = mesh
-        self.use_kernel = envs.USE_MOE_EP_KERNEL
+        self.use_kernel = envs.USE_MOE_EP_KERNEL and moe.use_ep
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/quantization/unquantized.py:171]: hm, if we don't use ep, why can't we use the fused moe kernel?

File: tpu_inference/layers/vllm/fused_moe.py (hunk 6 above)
Comment (vanbasten23) [reply tpu_inference/layers/vllm/fused_moe.py:419]: hmm, I ran `pytest -s -v tests/layers/vllm/test_unquantized.py -k test_fused_moe` but I couldn't trigger that branch, even with use_ep==True. In the code, use_ep is always false so that branch is never triggered. Am I missing something?

You can stick a print and try yourself.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 2 above)
Comment (vanbasten23) [reply tpu_inference/layers/vllm/fused_moe.py:129]: Thanks for the context.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 7 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/unquantized.py:231]: `jnp.transpose(w13_reshaped, (0, 1, 3, 2))` and `jnp.swapaxes(w13_reshaped, 2, 3)` are the same things. The latter is design specifically for only transposing two axes - which is what `jnp.transpose` does by going from `(0, 1, 2, 3)` to `(0, 1, 3, 2)`. So using `swapaxes` is simpler.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 8 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/unquantized.py:260]: So i've deleted that because it was misleading. That path does utilize kernel - it's just a different kernel.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 9 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/unquantized.py:171]: The fused moe kernel is designed specifically for EP use case. If you look at the kernel code, there is no code path to support TP.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 5 above)
Comment (vanbasten23) [reply tpu_inference/layers/vllm/fused_moe.py]: Could you add a comment please?

Also, here it seems that it changes the previous approach? Previously, send_sizes is a 1d array of size `ep_size`. But right now, since you don't specify the axis, the send_sizes will be a scalar.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 9 above)
Comment (yaochengji) [reply tpu_inference/layers/vllm/quantization/unquantized.py:171]: the fused moe kernel only supports ep at the moment.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 6 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py:419]: You were right! And I was also able to catch easy to miss bug after I fixed the unit test. Thanks! And I've modified the unit test code.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 5 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py]: Done. Also, your comment is correct and this was fixed after the unit test fix I've mentioned in another thread.

**Review by vanbasten23 — APPROVED** [review]
